packages/darklang/languageTools/lsp-server/fileSystemProvider.dark

```diff
@@ -0,0 +1,828 @@
+module Darklang =
+  module LanguageTools =
+    module LspServer =
+      module FileSystemProvider =
+        type FileType =
+          | File
+          | Directory
+          | SymbolicLink
+
+        type FileSystemErrorType =
+          | FileNotFound
+          | FileExists
+          | FileNotADirectory
+          | FileIsADirectory
+          | NoPermissions
+          | Unavailable
+          | Other
+
+        type FileSystemProviderClientCapabilities = { queryable: Bool }
+
+        type URI = String
+
+        module FileSystemProviderOptions =
+          type FileSystemProviderOptions =
+            { scheme: String
+              caseSensitive: Bool
+              readonly: Bool }
+
+          let toJson (options: FileSystemProviderOptions) : Json =
+            [ Stdlib.Option.Option.Some(("scheme", Json.String options.scheme))
+              Stdlib.Option.Option.Some(
+                ("caseSensitive", Json.Bool options.caseSensitive)
+              )
+              Stdlib.Option.Option.Some(("readonly", Json.Bool options.readonly)) ]
+            |> Stdlib.Option.values
+            |> Json.Object
+
+
+        module ReadFile =
+          let method = "fileSystem/read"
+
+          module ReadFileParams =
+            type ReadFileParams = { uri: URI }
+
+            let fromJson
+              (json: Json)
+              : Stdlib.Result.Result<ReadFileParams, String> =
+              match json with
+              | Object fields ->
+                match
+                  Stdlib.List.findFirst fields (fun (key, _) -> key == "uri")
+                with
+                | Some((_, String uri)) ->
+                  (ReadFileParams { uri = uri }) |> Stdlib.Result.Result.Ok
+                | _ -> Stdlib.Result.Result.Error "Expected uri"
+              | _ -> Stdlib.Result.Result.Error "Expected object"
+
+
+          module ReadFileResponse =
+            type ReadFileResponse = { content: String }
+
+            let toJson (r: ReadFileResponse) : Json =
+              [ Stdlib.Option.Option.Some(("content", Json.String r.content)) ]
+              |> Stdlib.Option.values
+              |> Json.Object
+
+
+          let handleReadFileRequest
+            (state: LspState)
+            (requestId: JsonRPC.RequestId)
+            (params: ReadFileParams.ReadFileParams)
+            : LspState =
+            // Remove the darklang:// prefix and the .dark suffix from the URI, (e.g.darklang://type/Darklang/Stdlib/Option/Option.dark)
```

Comment:
let's add a CLEANUP to make this 'remove scheme part of URI' a bit more universal.
In fact, something like that might be useful in our Stdlib -- like `Stdlib.Uri.removeScheme" or something.

Probably not worth any effort as part of this PR, but this feels a tad precarious as is, so worth the CLEANUP.